### PR TITLE
Ports changeling to role datums

### PR DESCRIPTION
**Pull request by MadmanMartian**
- [x] Fix the changeling vial
- [x] Figure out what we're doing with Topic()
- [x] Add Objectives

### CODE REVIEW

File: code/datums/gamemode/factions/faction.dm [hunk 1]
```diff
@@ -39,6 +39,7 @@
 	objective_holder = new
 	forgeObjectives()
 	for(var/datum/role/R in members)
+		to_chat(world, "Calling Role OnPostSetup at line 43")
```
Comment (MadmanMartian) [inline code/datums/gamemode/factions/faction.dm]: HITLERS

File: code/datums/gamemode/gamemode.dm [hunk 2]
```diff
@@ -83,6 +86,51 @@
...
+		//else
+			//Whichever witchcraft we employ in the future to have it scale with the playercount
+
+
+/datum/gamemode/proc/CreateNumOfRoles(var/datum/role/R, var/num)
+	var/list/available_players = get_ready_players()
+	for(var/mob/new_player/P in available_players)
+		if(!P.client || !P.mind)
+			available_players.Remove(P)
+			to_chat(world, "[P] had no mind or client.")
```
Comment (MadmanMartian) [inline code/datums/gamemode/gamemode.dm]: HITLERS

File: code/datums/gamemode/gamemode.dm [hunk 3]
```diff
@@ -83,6 +86,51 @@
...
+			continue
+		if(!P.client.desires_role(initial(R.required_pref)) || jobban_isbanned(P, initial(R.required_pref)))
+			available_players.Remove(P)
+			to_chat(world, "with required_pref being [initial(R.required_pref)], [P] was either job_banned ([jobban_isbanned(P, initial(R.required_pref))]) or did not desire the role ([!P.client.desires_role(initial(R.required_pref))])")
```
Comment (MadmanMartian) [inline code/datums/gamemode/gamemode.dm]: HITLERS

File: code/datums/gamemode/gamemode.dm [hunk 4]
```diff
@@ -105,6 +153,9 @@
 
 	for(var/datum/faction/F in factions)
 		F.OnPostSetup()
+	for(var/datum/role/R in orphaned_roles)
+		to_chat(world, "Calling Role OnPostSetup in gamemode.dm line 158")
```
Comment (MadmanMartian) [inline code/datums/gamemode/gamemode.dm]: HITLERS

File: code/datums/gamemode/role/changeling.dm [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+
+
+					body = "<table><tr><td>";
+					body +=	"</td><td align='center'>";
+					body +=	"<font size='2'><b>"+desc+"</b></font> <BR>";
+					body +=	"<font size='2'><font color = 'red'><b>"+helptext+"</b></font> <BR>";
+
+					if(!ownsthis)
+					{
+						body += "<a href='?src=\ref[src];P="+power+";mind=\ref[antag];'>Evolve</a>"
```
Comment (MadmanMartian) [inline code/datums/gamemode/role/changeling.dm:369]: a href='?src=\ref[src];mind=\ref[antag];(whatever else)' is how we should approach html links in the future involving role HTML stuff

File: code/datums/gamemode/role/role.dm [hunk 6]
```diff
@@ -130,8 +133,8 @@
 	if(faction && src in faction.members)
 		faction.members.Remove(src)
 
-	if(!faction)
-		ticker.mode.orphaned_roles -= src
+	if(!faction && src in ticker.mode.orphaned_roles)
```
Comment (DamianX) [inline code/datums/gamemode/role/role.dm]: this check you added is not needed: `Remove` will have to search through the list for `src` anyway

File: code/datums/gamemode/objectives/absorb.dm [hunk 7]
```diff
@@ -0,0 +1,15 @@
+/datum/objective/absorb
+	explanation_text = "Absorb a number of genomes."
+	var/genomes_to_absorb
+/datum/objective/absorb/New()
+	..()
+	genomes_to_absorb = rand(2,3)
+	explanation_text = "Absorb [genomes_to_absorb] genomes."
+
+/datum/objective/absorb/IsFulfilled()
+	..()
```
Comment (DamianX) [inline code/datums/gamemode/objectives/absorb.dm]: either `. = ..()` or don't call this at all, I guess?

File: code/datums/gamemode/role/changeling.dm [hunk 8]
```diff
@@ -0,0 +1,599 @@
...
+	var/datum/dna/chosen_dna
+	for(var/datum/dna/DNA in absorbed_dna)
+		if(dna_owner == DNA.real_name)
+			chosen_dna = DNA
+			break
+	return chosen_dna
+
+// READ: Don't use the apostrophe in name or desc. Causes script errors.
+
+var/list/powers = typesof(/datum/power/changeling) - /datum/power/changeling	//needed for the badmin verb for now
```
Comment (DamianX) [inline code/datums/gamemode/role/changeling.dm]: `subtypesof`

File: code/datums/gamemode/role/changeling.dm [hunk 9]
```diff
@@ -0,0 +1,599 @@
...
+	helptext = "The blade can be retracted by using the same verb used to manifest it. It has a chance to deflect projectiles."
+	genomecost = 5
+	verbpath = /obj/item/verbs/changeling/proc/changeling_armblade
+
+
+/datum/role/changeling/proc/EvolutionMenu()
+	set category = "Changeling"
+	set desc = "Level up!"
+
+	if(!usr || !usr.mind || !usr.mind.GetRole(CHANGELING))
```
Comment (DamianX) [inline code/datums/gamemode/role/changeling.dm]: not a big deal but you might want to only call `usr.mind.GetRole(CHANGELING)` once instead of twice

**Comment by Durgen** [comment]
Are changelings getting spells for verbs like vamps apparently are?

**Comment by MadmanMartian** [comment]
sometime in the future

File: code/game/objects/items/changeling_vial.dm [hunk 10]
```diff
@@ -7,26 +7,29 @@
...
-				M.make_new_changeling(0,1)
-				H.mind.changeling.geneticpoints = Clamp(genomes_to_give, 0, 100)
 				qdel(src)
 				H.sleeping += 10
 				sleep(100)
 				to_chat(H, "You feel your consciousness slipping away...")
 				sleep(100)
+				var/datum/role/changeling/C = new(M)
+				if(C)
+					C.geneticpoints+=genomes_to_give
```
Comment (gbasood) [inline code/game/objects/items/changeling_vial.dm:25]: btw you messed this up :^)
(this is the thing I'm fixing in the rebase)

Comment (MadmanMartian) [reply code/game/objects/items/changeling_vial.dm:25]: aw bugger

**Pull request merged**